tsl/src/remote/copy_fetcher.c

```diff
@@ -338,85 +383,102 @@ copy_fetcher_complete(CopyFetcher *fetcher)
+				 * check against somehow getting out of sync with the data.
+				 */
+				fetcher->file_trailer_received = true;
 
-			/*
-			 * There is also one case where no tupdesc attributes are retrieved.
-			 * This is when we do `select count(*) from t`, and
-			 * `enable_partitionwise_aggregate` is 0, so the data node queries
-			 * become `select null from ...` and we should get 1 NULL attribute
-			 * from COPY.
-			 */
-			int16 expected_natts = Max(1, retrieved_natts);
-			if (natts != expected_natts)
-			{
-				elog(ERROR,
-					 "wrong number of attributes for a COPY tuple: expected %d, got %d",
-					 expected_natts,
-					 natts);
+				/* Next PQgetCopyData() should return -1, indicating EOF and
+				 * that the remote side ended the copy. The final result
+				 * (PGRES_COMMAND_OK) should then be read with
+				 * PQgetResult(). */
 			}
-
-			Datum *values = &fetcher->batch_values[tupdesc_natts * row];
-			bool *nulls = &fetcher->batch_nulls[tupdesc_natts * row];
-			for (int i = 0; i < tupdesc_natts; i++)
+			else
 			{
-				nulls[i] = true;
-			}
+				Assert(!fetcher->file_trailer_received && !fetcher->state.eof);
+				/*
+				 * There is also one case where no tupdesc attributes are retrieved.
+				 * This is when we do `select count(*) from t`, and
+				 * `enable_partitionwise_aggregate` is 0, so the data node queries
+				 * become `select null from ...` and we should get 1 NULL attribute
+				 * from COPY.
+				 */
+				int16 expected_natts = Max(1, retrieved_natts);
+				if (natts != expected_natts)
+				{
+					elog(ERROR,
+						 "wrong number of attributes for a COPY tuple: expected %d, got %d",
+						 expected_natts,
+						 natts);
+				}
 
-			MemoryContextSwitchTo(fetcher->state.tuple_mctx);
-			for (int i = 0; i < retrieved_natts; i++)
-			{
-				const int att = list_nth_int(retrieved_attrs, i) - 1;
-				Assert(att >= 0);
-				Assert(att < tupdesc_natts);
-				const int32 att_bytes = copy_data_read_int32(&copy_data);
-				if (att_bytes == -1)
+				Datum *values = &fetcher->batch_values[tupdesc_natts * row];
+				bool *nulls = &fetcher->batch_nulls[tupdesc_natts * row];
+				for (int i = 0; i < tupdesc_natts; i++)
 				{
-					/*
-					 * NULL. From the Postgres docs:
-					 * Usually, a receive function should be declared STRICT; if
-					 * it is not, it will be called with a NULL first parameter
-					 * when reading a NULL input value. The function must still
-					 * return NULL in this case, unless it raises an error.
-					 * (This case is mainly meant to support domain receive
-					 * functions, which might need to reject NULL inputs.)
-					 * https://www.postgresql.org/docs/current/sql-createtype.html
-					 */
-					if (!attconv->conv_funcs[att].fn_strict)
-					{
-						values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
-														  NULL,
-														  attconv->ioparams[att],
-														  attconv->typmods[att]);
-					}
-					else
+					nulls[i] = true;
```

Comment:
`memset(nulls, 1, sizeof(nulls));` might be better than a loop

Reply:
This was old code, just indented differently. Also, looks like the loop is necessary anyway and I guess nulls was initialized to "false" so this is only called in the "true" case.  I will leave further changes out of this PR.